src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb

```diff
@@ -0,0 +1,373 @@
+                    expressionType = SyntaxKind.EqualsExpression
+                    operatorType = SyntaxKind.EqualsToken
+                End If
+
+                result = SyntaxFactory.BinaryExpression(
+                    expressionType,
+                    binaryExpression.Left.Parenthesize(),
+                    SyntaxFactory.Token(
+                        binaryExpression.OperatorToken.LeadingTrivia,
+                        operatorType,
+                        binaryExpression.OperatorToken.TrailingTrivia),
+                    binaryExpression.Right.Parenthesize())
+
+                result = result _
+                .WithLeadingTrivia(binaryExpression.GetLeadingTrivia()) _
+                .WithTrailingTrivia(binaryExpression.GetTrailingTrivia())
+
+                Return True
+            End If
+
+            Return False
+        End Function
+
+        Private Function TryNegateBinaryLogicalExpression(
+            expression As ExpressionSyntax,
+            semanticModel As SemanticModel,
+            cancellationToken As CancellationToken,
+            ByRef result As ExpressionSyntax) As Boolean
+
+            Dim inverses As Tuple(Of SyntaxKind, SyntaxKind) = Nothing
+            If s_logicalInversesMap.TryGetValue(expression.Kind, inverses) Then
+                Dim binaryExpression = DirectCast(expression, BinaryExpressionSyntax)
+
+                ' NOTE: result must be parenthesized because And & AndAlso have higher precedence than Or & OrElse
+                result = SyntaxFactory.BinaryExpression(
+                    inverses.Item1,
+                    Negate(binaryExpression.Left, semanticModel, cancellationToken),
+                    SyntaxFactory.Token(
+                        binaryExpression.OperatorToken.LeadingTrivia,
+                        inverses.Item2,
+                        binaryExpression.OperatorToken.TrailingTrivia),
+                    Negate(binaryExpression.Right, semanticModel, cancellationToken))
+
+                result = result _
+                    .Parenthesize() _
+                    .WithLeadingTrivia(binaryExpression.GetLeadingTrivia()) _
+                    .WithTrailingTrivia(binaryExpression.GetTrailingTrivia())
+
+                Return True
+            End If
+
+            Return False
+        End Function
+
+        Protected Function Negate(expression As ExpressionSyntax, semanticModel As SemanticModel, cancellationToken As CancellationToken) As ExpressionSyntax
+            Dim result As ExpressionSyntax = Nothing
+
+            If TryNegateBinaryComparisonExpression(expression, semanticModel, cancellationToken, result) OrElse
+               TryNegateBinaryLogicalExpression(expression, semanticModel, cancellationToken, result) Then
+                'Return result.WithAdditionalAnnotations(Formatter.Annotation)
+                Return result
+            End If
+
+            Select Case expression.Kind
```

Comment:
these cases seem very similar to the C# version.  Can we extract common code to the base class by using ISyntaxFacts?

Reply:
Note we woudl probably need to expand ISyntaxFacts a bit.  But i like tehn that that type gets more utility, and we remove more language specific stuff from our individual features.